### PR TITLE
[docs] Upgrade to Next.js 15

### PR DESCRIPTION
**Pull request by vladmoroz**
Upgrade to Next.js 15, ~mainly for the TypeScript config.~

~Since we are going to use built-in @next/mdx features that are configured through the Next.js config, we need this in order to use rehype/remark plugins written in TypeScript.~

Edit: forget this, `next.config.ts` module resolution is CJS only, and all rehype/remark ecosystem is ESM

*** 

I also briefly tried enabling Turbopack, but it doesn't work with the current demos which rely on Webpack to dynamically load the components. I'll migrate to Turbopack when we don't do the dynamic Webpack import anywhere.

### CODE REVIEW

File: docs/package.json [hunk 1]
```diff
@@ -70,8 +70,8 @@
     "@types/hast": "^3.0.4",
     "@types/node": "^18.19.54",
     "@types/prop-types": "^15.7.13",
-    "@types/react": "^18.3.10",
-    "@types/react-dom": "^18.3.0",
+    "@types/react": "npm:types-react@19.0.0-rc.1",
+    "@types/react-dom": "npm:types-react-dom@19.0.0-rc.1",
```
Comment (vladmoroz) [inline docs/package.json:74]: This comes from their codemod

File: docs/package.json [hunk 2]
```diff
@@ -39,14 +39,14 @@
...
     "postcss": "^8.4.47",
     "postcss-import": "^16.1.0",
     "prop-types": "^15.8.1",
-    "react": "^18.3.1",
-    "react-dom": "^18.3.1",
+    "react": "19.0.0-rc-69d4b800-20241021",
+    "react-dom": "19.0.0-rc-69d4b800-20241021",
     "react-error-boundary": "^4.0.13",
-    "react-is": "^18.3.1",
+    "react-is": "19.0.0-rc-69d4b800-20241021",
```
Comment (vladmoroz) [inline docs/package.json:49]: This comes from their codemod

File: docs/shiki.config.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+export const shikiTheme = {
+  light: 'github-light-default',
+  dark: 'material-theme-ocean',
+};
```
Comment (vladmoroz) [inline docs/shiki.config.ts:4]: Next.js was complaining about exporting a `config` from the project root, seems like this used to be some pattern that is now deprecated

For context, this file will go completely when I'm done with code blocks updates

**Comment by michaldudak** [comment]
I have a feeling that making the regression tests pass will require upgrading React in all our packages, including `test` and Base UI.

**Comment by vladmoroz** [comment]
> I have a feeling that making the regression tests pass will require upgrading React in all our packages, including test and Base UI.

@michaldudak upgraded to React 19 everywhere and the regression tests now pass

~Should I worry about this?~

Edit: never mind, it passes after the rebase

image

~Running `pnpm proptypes` locally doesn't produce any changes~

**Comment by michaldudak** [comment]
We can disable the static route indicator (https://nextjs.org/docs/app/api-reference/next-config-js/devIndicators#appisrstatus-static-indicator). All our routes are prerendered at build time, so it doesn't add any value.

**Comment by michaldudak** [comment]
I configured the CI to also run tests against React 18, as 19 won't be widely adopted yet

**Comment by vladmoroz** [comment]
> We can disable the static route indicator (https://nextjs.org/docs/app/api-reference/next-config-js/devIndicators#appisrstatus-static-indicator). All our routes are prerendered at build time, so it doesn't add any value.

@michaldudak nice, updated. Good to merge?

**Review by michaldudak — APPROVED** [review]

**Comment by michaldudak** [comment]
FYI, ESM support in next.config.ts is coming: https://github.com/vercel/next.js/pull/68365

**Pull request merged**